lib/trie/proof.go

```diff
@@ -19,126 +19,274 @@ package trie
+	entry := (*s)[len(*s)-1]
+	// removes the top of the stack
+	*s = (*s)[:len(*s)-1]
+	return entry
+}
 
-			length := lenCommonPrefix(n.key, key)
-			currNode = n.children[key[length]]
-			key = key[length+1:]
-		}
+// Last returns the top of the stack without remove from the stack
+func (s *Stack) Last() *StackEntry {
+	if len(*s) < 1 {
+		return nil
 	}
+	return (*s)[len(*s)-1]
+}
 
-	for _, n := range nodes {
-		var (
-			hashNode    []byte
-			encHashNode []byte
-			err         error
-		)
-
-		if encHashNode, hashNode, err = n.encodeAndHash(); err != nil {
-			return 0, fmt.Errorf("problems while encoding and hashing the node: %w", err)
-		}
+// StackIterator is used to transform the stack into a iterator, the iterator starts the cursor (index) from 0
+type StackIterator struct {
+	index int
+	set   []*StackEntry
+}
 
-		if err = db.Put(hashNode, encHashNode); err != nil {
-			return len(nodes), err
-		}
+// Next returns the current item the cursor is on and increment the cursor by 1
+func (i *StackIterator) Next() *StackEntry {
+	if i.HasNext() {
+		t := i.set[i.index]
+		i.index++
+		return t
 	}
 
-	return len(nodes), nil
+	return nil
 }
 
-// VerifyProof checks merkle proofs given an proof
-func VerifyProof(rootHash common.Hash, key []byte, db chaindb.Reader) (bool, error) {
-	key = keyToNibbles(key)
-	if len(key) == 0 {
-		return false, ErrEmptyNibbles
+// Peek returns the current item the cursor is on but dont increment the cursor by 1
+func (i *StackIterator) Peek() *StackEntry {
+	if i.HasNext() {
+		return i.set[i.index]
 	}
+	return nil
+}
 
-	var wantedHash []byte
-	wantedHash = rootHash.ToBytes()
+// HasNext returns true if there is more itens into the iterator
+func (i *StackIterator) HasNext() bool {
+	return i.index < len(i.set)
+}
 
-	for {
-		enc, err := db.Get(wantedHash)
-		if errors.Is(err, chaindb.ErrKeyNotFound) {
-			return false, nil
-		} else if err != nil {
-			return false, nil
-		}
+func (s *Stack) iter() *StackIterator {
+	iter := &StackIterator{index: 0}
+	iter.set = make([]*StackEntry, len(*s))
+	copy(iter.set, *s)
+
+	return iter
+}
+
+// Step struct helps the proof generation with the next step that should be taken when
+// the seek node is found or to keep searching from the next node
+// using the offset of the nibbled key
+type Step struct {
+	Found     bool
+	Value     []byte
+	NextNode  []byte
+	KeyOffset int
+}
+
+// GenerateProof receive the keys to proof, the trie root and a reference to database
+// will
+func GenerateProof(root []byte, keys [][]byte, db chaindb.Database) ([][]byte, error) {
+	stack := make(Stack, 0)
+	proofs := make([][]byte, 0)
+
+	for _, k := range keys {
+		nk := keyToNibbles(k)
+
+		unwindStack(&stack, proofs, nk)
 
-		currNode, err := decodeBytes(enc)
+		lookup := NewLookup(root, db)
+		recorder := new(Recorder)
+		expectedValue, err := lookup.Find(nk, recorder)
 		if err != nil {
-			return false, fmt.Errorf("could not decode node bytes: %w", err)
+			return nil, err
+		}
```

Comment:
the value isn't required for the proof. I'm also not sure what the purpose of the code after here is, as when I tried it previously the proof was already in the `recorder` at this point

Reply:
you're right, the proof is inside recorder, I remove the unecessary code and I added a map of `trackedProofs` to dont duplicate already added proofs